source/develop/release-management/features/external-network-provider.html.md

```diff
@@ -0,0 +1,525 @@
+Called before/after a nic is hot plugged.
+Network provider reponsibilities:
+
+* make sure the network exists on this host
+
+* connect the VM NIC to the specified external network
+
+#### before_nic_hotunplug
+Called before a nic is hot unplugged.
+Network provider reponsibilities:
+
+* unplug the VM NIC from the external network (if required)
+
+#### before_vm_migration_source
+Called on the source host before a VM is migrated.
+Network provider reponsibilities:
+
+* unplug the VM NIC from the external network (if required)
+
+#### before_vm_migration_destination
+Called on the destination host before a VM is migrated.
+Network provider reponsibilities:
+
+* make sure the network exists on this host
+
+* connect the VM NIC to the specified external network
+
+#### after_get_stats
+Called after generating the network statistics which are to be send to the engine
+Network provider reponsibilities:
+
+* provide the statistics for the VM NIC attached to the external network
+
+#### after_get_caps
+Called after generating the host capabilites (host configuration) which are to be send to the engine.
+Network provider reponsibilities:
+
+* provider information about the host network configuration related to the external network.
+Depending on the implementation, the provider might want report the external network being
+attached to any of the host NICS, report the external network being attached to a virtual host
+NIC, hide some host NICs used by the external network implementation, or report any other implemented configuration.
+
+#### before_network_setup/after_network_setup
+Called before and after the network setup operation on the host.
+Network provider reponsibilities:
+
+* if needed, the network provider might prevent network setup actions which would interfere with the external networks on the host.
+
+
+## External provider reference implementation
+
+A reference implementation of an external provider is prepared to show how.
+The implementation show in the reference implementation is very simple (naive),
+it's purpose is to show where rather than how different parts of the
+functionality should be implemented.
+
+The reference implementation consists of two parts:
+
+* the REST API - implementing the external provider REST API
+
+* VIF driver - implementing the driver which connects VM NICs on the hosts
+
+
+
+## Features under discussion
+
+### Provider custom properties
+
+Custom properties for an external network provider will be added. The user will be able to add the custom properties when adding the provider. The changed UI will should look like this (best guess - this might still change):
+
+![](external_network_provider_custom_properties.png "fig:external_network_provider_custom_properties.png")
+
+These properties will be passed to the VIF driver upon pluging of a nic, the same way the "vnic_id" property
+is passed.
+These properties might also be passed to the VIF driver when unplugging a nic (to be decided).
```

Comment:
This would be one of many examples of what one can do with the properties. I modified this to:

As an example, this would allow to add custom properties to native networks vNIC profiles. vNIC profiles are not available for external network; this subfeature would allow another means to pass provider-specific customization to starting up a vNIC.